### PR TITLE
Bank Tags: refactor to use new ChatboxPanelManager

### PR DESCRIPTION
**Pull request by raiyni**
Closes https://github.com/runelite/runelite/issues/5892
Closes https://github.com/runelite/runelite/issues/1297
Closes https://github.com/runelite/runelite/issues/1394

### CODE REVIEW

**Comment by deathbeam** [comment]
Does this also close the bank tags related bugs?

**Comment by raiyni** [comment]
Yes, updated description

File: runelite-api/src/main/java/net/runelite/api/ScriptEvent.java [hunk 1]
```diff
@@ -35,5 +37,11 @@
 	int WIDGET_TARGET_INDEX = -2147483641;
 	int KEY_CODE = -2147483640;
 	int KEY_CHAR = -2147483639;
-	String NAME =  "event_opbase";
+	String NAME = "event_opbase";
+
+	Widget getSource();
```
Comment (deathbeam) [inline runelite-api/src/main/java/net/runelite/api/ScriptEvent.java]: You should probably document all 3 methods

File: runelite-client/src/main/java/net/runelite/client/plugins/banktags/tabs/TabInterface.java [hunk 2]
```diff
@@ -922,6 +917,39 @@ private void updateWidget(Widget t, int y)
...
+	{
+		// In case the widget ids array is incorrect, do not proceed
+		if (widgetIds == null || widgetIds.length < 21)
+		{
+			return;
+		}
+
+		clientThread.invoke(() ->
+		{
+
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/banktags/tabs/TabInterface.java]: remove this empty line

File: runelite-client/src/main/java/net/runelite/client/plugins/banktags/tabs/TabInterface.java [hunk 3]
```diff
@@ -922,6 +917,39 @@ private void updateWidget(Widget t, int y)
 		t.revalidate();
 	}
 
+
+	public void fastSearch(InputType inputType, String search)
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/banktags/tabs/TabInterface.java]: what is the difference between this method and doSearch method?

File: runelite-client/src/main/java/net/runelite/client/plugins/banktags/tabs/TabInterface.java [hunk 4]
```diff
@@ -679,7 +674,7 @@ else if (entry.getOption().equals(SCROLL_DOWN))
 		}
 	}
 
-	private void resetSearch()
+	public void resetSearch()
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/banktags/tabs/TabInterface.java]: this can stay private

File: runelite-client/src/main/java/net/runelite/client/plugins/banktags/tabs/TabInterface.java [hunk 5]
```diff
@@ -218,6 +223,194 @@ public void init()
...
+
+							updateTabIfActive(tags);
+						});
+					})
+					.build();
+			});
+		}
+	}
+
+	private void handleNewTab(ScriptEvent event)
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/banktags/tabs/TabInterface.java]: maybe add some constants for all these event.getOp checks?

File: runelite-client/src/main/java/net/runelite/client/plugins/banktags/tabs/TabInterface.java [hunk 6]
```diff
@@ -218,6 +223,194 @@ public void init()
...
+		{
+			if (activeTab != null && Text.removeTags(event.getOpbase()).equals(activeTab.getTag()))
+			{
+				for (Integer item : items)
+				{
+					tagManager.addTag(item, activeTab.getTag());
+				}
+
+				openTag(TAG_SEARCH + activeTab.getTag());
+			}
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/banktags/tabs/TabInterface.java]: do early return here instead of creating else block

File: runelite-client/src/main/java/net/runelite/client/plugins/banktags/tabs/TabInterface.java [hunk 7]
```diff
@@ -218,6 +223,194 @@ public void init()
...
+		}
+		else
+		{
+			clientThread.invokeLater(() ->
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/banktags/tabs/TabInterface.java]: why this needs to be in clientThread invokeLater?

File: runelite-client/src/main/java/net/runelite/client/plugins/banktags/BankTagsPlugin.java [hunk 8]
```diff
@@ -252,22 +253,33 @@ public void onMenuOptionClicked(MenuOptionClicked event)
...
-				// the tag got added in new tags
-				final List<String> initialTags = SPLITTER.splitToList(initialValue.toLowerCase());
-				final List<String> tags = SPLITTER.splitToList(newTags.toLowerCase());
-				tabInterface.updateTabIfActive(initialTags);
-				tabInterface.updateTabIfActive(tags);
-			});
+			chatboxPanelManager.openTextInput(itemName + " tags:")
+				.value(initialValue)
+				.onDone((newTags) ->
+				{
```
Comment (Abextm) [inline runelite-client/src/main/java/net/runelite/client/plugins/banktags/BankTagsPlugin.java]: nit: these braces are unnecessary and so are the parens around `newTags`

File: runelite-client/src/main/java/net/runelite/client/plugins/banktags/tabs/MenuIndexes.java [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+
+import lombok.AllArgsConstructor;
+import lombok.Getter;
+
+class MenuIndexes
+{
+	@AllArgsConstructor
+	enum NewTab
+	{
+		NEW_TAB(2),
```
Comment (Abextm) [inline runelite-client/src/main/java/net/runelite/client/plugins/banktags/tabs/MenuIndexes.java]: These can just be `public static final int`. They gain nothing from being types.

File: runelite-client/src/main/java/net/runelite/client/plugins/banktags/tabs/TabInterface.java [hunk 10]
```diff
@@ -922,19 +918,23 @@ private void updateWidget(Widget t, int y)
 		t.revalidate();
 	}
 
-	private void doSearch(InputType inputType, String search)
+
```
Comment (Abextm) [inline runelite-client/src/main/java/net/runelite/client/plugins/banktags/tabs/TabInterface.java]: remove this whitespace

File: runelite-client/src/main/java/net/runelite/client/plugins/banktags/tabs/MenuIndexes.java [hunk 11]
```diff
@@ -25,59 +25,20 @@
...
 package net.runelite.client.plugins.banktags.tabs;
 
-import lombok.AllArgsConstructor;
-import lombok.Getter;
-
 class MenuIndexes
 {
-	@AllArgsConstructor
-	enum NewTab
+	static class NEWTAB
```
Comment (Abextm) [inline runelite-client/src/main/java/net/runelite/client/plugins/banktags/tabs/MenuIndexes.java]: class name stays CamelCase

File: runelite-client/src/main/java/net/runelite/client/plugins/banktags/tabs/TabInterface.java [hunk 12]
```diff
@@ -732,8 +724,11 @@ private void deleteTab(String tag)
 		configManager.unsetConfiguration(CONFIG_GROUP, ICON_SEARCH + tag);
 		tabManager.save();
 
-		updateBounds();
-		scrollTab(0);
+		clientThread.invoke(() ->
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/banktags/tabs/TabInterface.java]: You already have clientThread.invoke in place where you call deteleTab, why not just put entire deleteTab call into clientThread? And remove this invoke here

File: runelite-client/src/main/java/net/runelite/client/plugins/banktags/BankTagsPlugin.java [hunk 13]
```diff
@@ -252,22 +253,30 @@ public void onMenuOptionClicked(MenuOptionClicked event)
...
+						// Check both previous and current tags in case the tag got removed in new tags or in case
+						// the tag got added in new tags
+						final List<String> initialTags = SPLITTER.splitToList(initialValue.toLowerCase());
+						final List<String> tags = SPLITTER.splitToList(newTags.toLowerCase());
+						tabInterface.updateTabIfActive(initialTags);
+						tabInterface.updateTabIfActive(tags);
+					}))
+				.build();
+
+			if (isSearchOpen)
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/banktags/BankTagsPlugin.java]: Why is this necessary now?

Comment (deathbeam) [reply runelite-client/src/main/java/net/runelite/client/plugins/banktags/BankTagsPlugin.java]: You have bad habit of putting everything to 1 PR without explanation. You should at least split that to separate commit and reference that issue

File: runelite-client/src/main/java/net/runelite/client/plugins/banktags/tabs/TabInterface.java (hunk 12 above)
Comment (deathbeam) [reply runelite-client/src/main/java/net/runelite/client/plugins/banktags/tabs/TabInterface.java]: Yeah, and in both cases it is from the ChatBoxPanelManager. Imo it would be better to wrap it in ClientThread there

File: runelite-client/src/main/java/net/runelite/client/plugins/banktags/tabs/TabInterface.java [hunk 14]
```diff
@@ -191,7 +179,7 @@ public void init()
 		}
 
 		Widget bankContainer = client.getWidget(WidgetInfo.BANK_CONTAINER);
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/banktags/tabs/TabInterface.java]: You can probably just get the widgetIds directly from BANK_CONTAINER widget in the BankSearch when calling the method (getting widget has almost no overhead) instead of setting them here

Comment (raiyni) [reply runelite-client/src/main/java/net/runelite/client/plugins/banktags/tabs/TabInterface.java]: Rebased and moved.

File: runelite-client/src/main/java/net/runelite/client/plugins/banktags/tabs/BankSearch.java [hunk 15]
```diff
@@ -0,0 +1,115 @@
...
+	private static final int SEARCH_BUTTON_BACKGROUND_IDX = 15;
+	private static final int TITLE_BAR_IDX = 16;
+	private static final int ITEM_COUNT_IDX = 17;
+	private static final int TAB_BAR_IDX = 18;
+	private static final int INCINERATOR_IDX = 19;
+	private static final int INCINERATOR_CONFIRM_IDX = 20;
+	private static final int HIDDEN_WIDGET_IDX = 21;
+
+	@Inject
+	BankSearch(
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/banktags/tabs/BankSearch.java]: Make constructor private

File: runelite-client/src/main/java/net/runelite/client/plugins/banktags/tabs/BankSearch.java [hunk 16]
```diff
@@ -0,0 +1,115 @@
...
+import net.runelite.api.Client;
+import net.runelite.api.ScriptID;
+import net.runelite.api.VarClientInt;
+import net.runelite.api.VarClientStr;
+import net.runelite.api.vars.InputType;
+import net.runelite.api.widgets.Widget;
+import net.runelite.api.widgets.WidgetInfo;
+import net.runelite.client.callback.ClientThread;
+
+public class BankSearch
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/banktags/tabs/BankSearch.java]: Remove public modifier

File: runelite-client/src/main/java/net/runelite/client/plugins/banktags/tabs/BankSearch.java [hunk 17]
```diff
@@ -0,0 +1,115 @@
...
+{
+	private final Client client;
+	private final ClientThread clientThread;
+
+	// Widget indexes for searching
+	private static final int INNER_CONTAINER_IDX = 2;
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/banktags/tabs/BankSearch.java]: Move static fields above instance fields

File: runelite-client/src/main/java/net/runelite/client/plugins/banktags/tabs/BankSearch.java (hunk 16 above)
Comment (raiyni) [reply runelite-client/src/main/java/net/runelite/client/plugins/banktags/tabs/BankSearch.java]: needs to be public because it's used in `banktags` and `banktags.tabs`

Comment (deathbeam) [reply runelite-client/src/main/java/net/runelite/client/plugins/banktags/tabs/BankSearch.java]: Ah, right
